src/SelfServe/SelfServeTelemetryProcessor.ts

```diff
@@ -0,0 +1,124 @@
+import { sendMessage } from "../Common/MessageHandler";
+import { configContext } from "../ConfigContext";
+import { SelfServeMessageTypes } from "../Contracts/SelfServeContracts";
+import { appInsights } from "../Shared/appInsights";
+import { Action, ActionModifiers } from "../Shared/Telemetry/TelemetryConstants";
+import { userContext } from "../UserContext";
+import { SelfServeTelemetryMessage } from "./SelfServeTypes";
+
+export function trace(action: Action, data: SelfServeTelemetryMessage): void {
+  const actionModifier = ActionModifiers.Mark;
+  sendMessage({
+    type: SelfServeMessageTypes.TelemetryInfo,
+    data: {
+      action: Action[action],
+      actionModifier: actionModifier,
+      data: JSON.stringify(decorateData(data)),
+    },
+  });
+
+  appInsights.trackEvent({ name: Action[action] }, decorateData(data, actionModifier));
+}
+
+export function traceStart(action: Action, data: SelfServeTelemetryMessage): number {
+  const timestamp: number = Date.now();
+  sendMessage({
+    type: SelfServeMessageTypes.TelemetryInfo,
+    data: {
+      action: Action[action],
+      actionModifier: ActionModifiers.Start,
+      timestamp: timestamp,
+      data: JSON.stringify(decorateData(data)),
+    },
+  });
+
+  appInsights.startTrackEvent(Action[action]);
+  return timestamp;
+}
+
+export function traceSuccess(action: Action, data: SelfServeTelemetryMessage, timestamp?: number): void {
```

Comment:
These functions all look very similar. There seem to be a 1:1 mapping between `Action` and `ActionModifiers`. Consider reducing the code duplication by defining one single function (if it makes sense). Maybe you can take advantage of the `switch` that calls them. Or use a lookup table to get the right `ActionModifiers` (or use `ActionModifiers` directly if `Action` is redundant).